File: plugin-sdk-ts/src/profiles.ts
/**
 * Environment profiles module.
 *
 * Manage environment configuration snapshots for quick switching.
 *
 * Requires: `profiles_read` and/or `profiles_write` permissions.
 */
import { callHostJson } from './host';
import type { Profile, ProfileCreateInput } from './types';

/** List all profiles. Requires: profiles_read */
export function list(): Profile[] {
  return callHostJson<Profile[]>('cognia_profile_list', '');
}

/** Get a profile by ID. Requires: profiles_read */
export function get(id: string): Profile | null {
  return callHostJson<Profile | null>(
    'cognia_profile_get',
    JSON.stringify({ id }),
  );
}

/** Create a profile from the current environment state. Requires: profiles_write */
export function createFromCurrent(name: string, description?: string): string {
  return callHostJson<string>(
    'cognia_profile_create_from_current',
    JSON.stringify({ name, description: description ?? null }),
  );
}

/** Create a profile with explicit entries. Requires: profiles_write */
export function create(profile: ProfileCreateInput): string {
  return callHostJson<string>(
    'cognia_profile_create',
    JSON.stringify(profile),
  );
}

/** Apply (activate) a profile. Requires: profiles_write */
export function apply(id: string): void {
  callHostJson<{ ok: boolean }>(
    'cognia_profile_apply',
    JSON.stringify({ id }),
  );
}

/** Export a profile as JSON string. Requires: profiles_read */
export function exportProfile(id: string): string {
  return callHostJson<string>(
    'cognia_profile_export',
    JSON.stringify({ id }),
  );
}

/** Import a profile from JSON string. Requires: profiles_write */
export function importProfile(json: string): string {
  return callHostJson<string>(
    'cognia_profile_import',
    JSON.stringify({ json }),
  );
}

File: plugin-sdk/src/profiles.rs
//! Environment profiles module.
//!
//! Manage environment configuration snapshots for quick switching.

use crate::host;
use crate::types::*;
use extism_pdk::*;

/// List all profiles. Requires: profiles_read
pub fn list() -> Result<Vec<Profile>, Error> {
    let result = unsafe { host::cognia_profile_list(String::new())? };
    Ok(serde_json::from_str(&result)?)
}

/// Get a profile by ID. Requires: profiles_read
pub fn get(id: &str) -> Result<Option<Profile>, Error> {
    let input = serde_json::json!({ "id": id }).to_string();
    let result = unsafe { host::cognia_profile_get(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Create a profile from the current environment state. Returns profile ID. Requires: profiles_write
pub fn create_from_current(name: &str, description: Option<&str>) -> Result<String, Error> {
    let input = serde_json::json!({ "name": name, "description": description }).to_string();
    let result = unsafe { host::cognia_profile_create_from_current(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Create a profile with explicit entries. Returns profile ID. Requires: profiles_write
pub fn create(profile: &ProfileCreateInput) -> Result<String, Error> {
    let input = serde_json::to_string(profile).map_err(|e| Error::msg(e.to_string()))?;
    let result = unsafe { host::cognia_profile_create(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Apply (activate) a profile. Requires: profiles_write
pub fn apply(id: &str) -> Result<(), Error> {
    let input = serde_json::json!({ "id": id }).to_string();
    unsafe { host::cognia_profile_apply(input)?; }
    Ok(())
}

/// Export a profile as JSON string. Requires: profiles_read
pub fn export_profile(id: &str) -> Result<String, Error> {
    let input = serde_json::json!({ "id": id }).to_string();
    let result = unsafe { host::cognia_profile_export(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Import a profile from JSON string. Returns profile ID. Requires: profiles_write
pub fn import_profile(json: &str) -> Result<String, Error> {
    let input = serde_json::json!({ "json": json }).to_string();
    let result = unsafe { host::cognia_profile_import(input)? };
    Ok(serde_json::from_str(&result)?)
}
